### PR TITLE
Include block: trigger warning on incompatible context

### PR DESCRIPTION
**Pull request by matej21**
ref #60

### CODE REVIEW

**Comment by dg** [comment]
:+1:

File: tests/Latte/macros.include.escape.phpt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+
+$latte = new Latte\Engine;
+$latte->setTempDirectory(TEMP_DIR);
+$latte->setLoader(new \Latte\Loaders\StringLoader());
+
+Assert::error(function () use ($latte) {
+	$latte->renderToString('<meta content="{include #foo}">{block #foo}{$value}{/block}', array('value' => 'b"ar'));
+}, E_USER_WARNING, 'Incompatible context for include, use either |escape or |noescape modifier');
+
```
Comment (dg) [inline tests/Latte/macros.include.escape.phpt]: I think that "use either |escape or |noescape modifier" is misleading.

Comment (matej21) [inline tests/Latte/macros.include.escape.phpt]: any better idea?

File: src/Latte/Macros/BlockMacros.php [hunk 2]
```diff
@@ -125,6 +129,13 @@ public function macroInclude(MacroNode $node, PhpWriter $writer)
 		} else {
 			$cmd = 'Latte\Macros\BlockMacrosRuntime::callBlock' . ($parent ? 'Parent' : '') . "(\$_b, $name, %node.array? + " . ($parent ? 'get_defined_vars' : '$template->getParameters') . '())';
 		}
+		$context = $this->getCompiler()->getContext() ? : array(NULL, NULL);
```
Comment (matej21) [inline src/Latte/Macros/BlockMacros.php]: I'm not sure about this. Is there any difference between `getContext() = NULL` and `getContext() = array(NULL, NULL)`?

File: tests/Latte/macros.include.escape.phpt (hunk 1 above)
Comment (dg) [inline tests/Latte/macros.include.escape.phpt]: `'Incompatible context for include'` seems fine.

Comment (matej21) [inline tests/Latte/macros.include.escape.phpt]: I think it should say how to fix that.

Comment (dg) [inline tests/Latte/macros.include.escape.phpt]: How to fix it or how to bypass it?

Comment (matej21) [inline tests/Latte/macros.include.escape.phpt]: I don't think it as a bypass. Latte just does not know how to handle it properly, so it leaves the decision to the developer.

**Comment by matej21** [comment]
updated, squashed. I think it is ready to merge.

**Comment by dg** [comment]
IMHO `{block double double}` is not very intuitive, what about `{block "double"}?

**Comment by matej21** [comment]
that is not very explicit. usually, you don't name a block `double` but sth like `{block foo double}`  :)

**Comment by dg** [comment]
Of course something like `{block "foo"}`

**Comment by matej21** [comment]
I don't agree. I think we should use explicit syntax. Your version does not explicitly indicate that it somehow change the behaviour. It just seems like a quoted string. And some other developer unaware of this behaviour may remove it thinking that it is useless.

**Comment by dg** [comment]
What do you don't agree with? I am not suggesting that `{block "foo"}` is good solution, it is just an idea. My point is that relation between word `double` and HTML quotes is not predictable.

**Comment by matej21** [comment]
This feature will not be used very often so I think we should be as explicit as possible. Maybe `{block foo doubleQuotes}`?

**Comment by dg** [comment]
This is little bit complicated… because context is not only about escaping, it is about parsing too. The workaround is

```
<foo bar="{block foo}…{/block}">
```

File: src/Latte/Macros/BlockMacros.php [hunk 3]
```diff
@@ -71,6 +74,7 @@ public function finalize()
 				$func = '_lb' . substr(md5($this->getCompiler()->getTemplateId() . $name), 0, 10) . '_' . preg_replace('#[^a-z0-9_]#i', '_', $name);
 				$snippet = $name[0] === '_';
 				$prolog[] = "//\n// block $name\n//\n"
+					. "\$_b->types[" . var_export($name, TRUE) . "][] = " . var_export($this->blockTypes[$name], TRUE) . ";"
```
Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: The `[]` means, that blocks with same name can have a different contexts? Is it reasonable?

**Comment by dg** [comment]
Btw, I think we can always escape both quotes and remove difference between single / double / no quotes.

File: src/Latte/Macros/BlockMacros.php (hunk 3 above)
Comment (matej21) [inline src/Latte/Macros/BlockMacros.php]: It just reflects blocks, which also use `[]`. I'm not really sure why :)

Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: Blocks have more definitions (more functions), but IMHO it should have only one type.

Comment (matej21) [inline src/Latte/Macros/BlockMacros.php]: Ok, it is probably needless. I will update it and will try do find some failing test :)

Comment (matej21) [inline src/Latte/Macros/BlockMacros.php]: Cannot be there any issues with dynamic blocks and/or overriding block from a parent template?

Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: Overriding block must have the same context as overrides, I think there is no issue…

**Comment by matej21** [comment]
rebased and partially rewritten. ad block overriding - there actually was an issue, but it should be fixed now :)

I've removed that second commit. Your solution/workaround is better. And in the future we can add sth like `n:omit` which will just extract a block and omit surrounding HTML.

**Comment by dg** [comment]
Thanks, I'll merge it today

**Comment by matej21** [comment]
Ok, I'll rebase once again (in an hour or so) and will update commit message and test names. Yesterday I was too focused on the functionality :)

**Comment by matej21** [comment]
Well.. with recent changes it has to be rewritten (almost) completely.

**Comment by dg** [comment]
What about something like this? https://github.com/dg/latte/commits/block-type

**Comment by matej21** [comment]
@dg that looks great!

**Comment by dg** [comment]
Feel free to take over and continue

**Comment by matej21** [comment]
rebased, renamed tests

File: src/Latte/Macros/BlockMacros.php [hunk 4]
```diff
@@ -123,6 +127,13 @@ public function macroInclude(MacroNode $node, PhpWriter $writer)
 			$cmd = 'Latte\Macros\BlockMacrosRuntime::callBlock' . ($parent ? 'Parent' : '') . "(\$_b, $name, %node.array? + " . ($parent ? 'get_defined_vars()' : '$this->params') . ')';
 		}
 
+		$node->modifiers = preg_replace('#\|noescape\s?(?=\||\z)#i', '', $node->modifiers, -1, $found);
+		if (!$found && !preg_match('#\|?escape(?:html|htmlcomment|ical|js|url|xml)?\s?(?=\||\z)#i', $node->modifiers)) {
+			$cmd = "if (" . var_export($this->exportBlockType(), TRUE) . " !== \$_b->types[$name]) { "
+				. "trigger_error('Incompatible context for include block $destination.', E_USER_WARNING); }\n"
```
Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: This is my fault, but maybe `for including` or `to include` is better.

**Comment by dg** [comment]
There is the last thing I am not sure. `|noescape` is here used as `|nocontextcheck`, right? Isn't it confusing? I removed it from exception message because it is not good idea to advise people to use `noescape` at the moment they do not know what is happening.

**Comment by matej21** [comment]
> There is the last thing I am not sure. `|noescape` is here used as `|nocontextcheck`, right? 

yes, that is right. 

> I removed it from exception message...

maybe we can write there sth like "escape it or define it in proper context" ?

**Comment by dg** [comment]
"define it in proper context" is good, but little bit obvious, i.e. "exception: block is in incompatible context" and "exception: block is in incompatible context, define it in proper context" give the same information.

"escape it" what do you mean?

**Comment by dg** [comment]
what about `|nowarning`? I'll can use it in `{foreach}` too.

**Comment by matej21** [comment]
By "escape" I mean "use `|escape` filter"..

noescape was really misleading, nocheck is ok, but nowarning was imho better :)

**Comment by dg** [comment]
I thinking about use nocheck as alias for (very bad name) nosafeurl, thus nocheck is better than nowarning.

**Comment by matej21** [comment]
ok, I will use |nocheck. What about the error message? should I include fixing possibilities (escape, nocheck) or the current message is enough?

**Comment by dg** [comment]
How to solve context incompatibility is really complex issue, so IMHO it will be better to use simple exception message and let user to google  doc/forum…

**Comment by matej21** [comment]
updated and added missing https://github.com/nette/latte/pull/65/files#diff-232e670f239e959d76e63b24a5bc9cb2R54

**Comment by dg** [comment]
Great! Can I merge it?

**Comment by matej21** [comment]
I think you can.

**Comment by matej21** [comment]
I'm an idiot. I forgot to change noescape to nocheck :)

**Comment by dg** [comment]
I'll fix it

**Comment by dg** [comment]
There is problem. For `<h1 n:block=title>` is context `html, tag`, because block is defined inside tag.

**Comment by matej21** [comment]
ok. should I send PR?

**Comment by dg** [comment]
Done

**Comment by matej21** [comment]
You should also check script/style tag, which has different content type: https://github.com/nette/latte/blob/master/src%2FLatte%2FCompiler.php#L359-L363

**Comment by dg** [comment]
Uff, it is really hard

**Pull request merged**